### ISSUE TITLE
The Big Rework

### ISSUE THREAD

**Issue by DavJCosby** [DavJCosby/sled#2]
This project is pretty messy. You can definitely tell that this was my first big project written in rust.

Here's a summary of some of the problems I have with the way this is written:
- `Room` is in charge of way more than it should be. It stores information about the room, bakes the position/angles of lights for faster calculations, manages multithreading of input and output devices, passes read/write-safe handles to those devices within those threads, etc etc.
- The thread-safety "solutions" used here are gross. At the time, the idea of using a `RWLock` to allow read-only access to `OutputDevice`'s and Read/Write access to `InputDevice`'s made sense, but now I see just how much spaghetti it introduces to the code.
- The `InputDevice`/`OutputDevice` model add so much unnecessary boilerplate and bulk.
  - Why do they even need to be structs in the first place?
  - Why do we need an output device? Wouldn't it be better to just pipe LED data back to the main thread or an output stream?
  - Why should the user be in charge of getting read/write permissions from the input/output handle?
- Big one: sliders aren't even a first class feature. Once you start an effect, there's no clean way to change the settings of that effect from the main thread. Requires the user to do some pretty hacky witchcraft to get it to play well with the thread logic.
- In general the file organization is pretty wack.

I'll write more specifics on what the new API will look like once I've had enough time to prototype my ideas, but for now here's the basic plan:
- `InputDevice` gets replaced by a simple closure
- `OutputDevice` gets phased out entirely and we instead take one of the following approaches:
  1. The user provides the SLED an output stream to dump LED values into, or
  2. LED values are returned by the same function call that generate the next frame of LEDs.
- The user ultimately decides how frequently to ask for new frames, not the InputDevice. In return, they get the same amount of frames back of data as they request, and at the exact same rate.
- Computation for input happens on the calling thread. At times we may briefly branch off into other threads for asynchronous work, and then quickly merge back to the main thread. For example, we could rewrite `map_angle_to_color()` to use a threadpool under the hood.
- Sliders will actually exist this time, and they will be owned and managed by the SLED. Sliders get passed in to the closure.
- Time information also gets passed in to the closure.
- More human-readable configuration files with added support for per-segment LED density.
- Big terminology changes all around. We're not calling it a "room" anymore, because we're trying to build a more general purpose statement.

Instead of reworking the existing pasta, I'll be building from the ground up on the [0.2.0-proto](https://github.com/DavJCosby/slc/tree/0.2.0-proto) branch.

**Comment by DavJCosby** [comment]
I've decided we can decouple some of the complexity here further. I'm breaking the responsibilities of SLED into two smaller components: `SLED` and `Driver.`

I realized that a SLED struct should serve as a drawable LED canvas, and that it didn't make much sense for a canvas to take ownership over the task of using itself. The previous plan looked something like this:

```rs
let sled = Sled::new("path/to/config.toml");

sled.set_driver(|canvas, sliders, time_info| {
  canvas.set_closest_to_point(
    (0.0, 0.0),
      sliders.get("foreground")?
  );

  Ok(()) 
});

let updated_colors = sled.step();
```

Though this is way better than what we were doing before, it just hasn't been sitting right with me. I realized that a driver needs to be external to the sled. For simple one-shot draw commands, no driver is necessary. Im still working out the full scope of the driver's responsibilities, but for now I'm thinking of it as a programmable "bot" that allows you to automate draw commands.

Here are a few examples of what I'm shooting for:
## One-Shot System, No Driver Necessary
```rs
let sled = Sled::new("path/to/config.toml");

sled.set_closest_to_point(
     (0.0, 0.0),
     Srgb::new(1.0, 1.0, 1.0)
);

let updated_colors = sled.get_colors();
```

## Run at a Fixed 60 fps
```rs
let sled = Sled::new("path/to/config/toml");
let driver = Driver::new();

driver.set_update_timing(UPDATE_TIMING::FIXED(60.0));
driver.set_commands(|sled, sliders, time_info| {
     let brightness = time_info.elapsed().sin() / 2.0 + 0.5;
     sled.set_all(Srgb::new(brightness, 0.0, 0.0));
});

driver.mount(sled);
loop {
     let updated_colors = driver.output_stream().read_frame()?;
}
```
The moment a driver with a fixed update schedule mounts the sled, updates begin. You can stop them by calling `.unmount()`.
## Manual Update Timing
```rs
let sled = Sled::new("path/to/config/toml");
let driver = Driver::new();
// by default, timing is set to UPDATE_TIMING::MANUAL;
driver.set_commands(|sled, sliders, time_info| {/* -snip- */});
driver.mount(sled);

driver.update();
// do some stuff...
driver.update();

let first_frame = driver.output_stream().read_frame()?;
let second_frame = driver.output_stream().read_frame()?;

```

## Passing Sled Ownership
Though in most cases it's best to just swap out the driver's commands, there are a few situations where you'd rather keep the drivers' settings intact. The `.mount(sled)` method grants the driver ownership over a sled, so getting that sled back is as simple as asking the driver to get off the thing.
```rs
// assuming we've already initialized a sled as well as driver1 and driver2..
driver1.mount(sled);

for i in 0..10 {
     driver1.update();
}

sled = driver1.unmount();
driver2.mount(sled);
```

# Final Thoughts
There are a few things still subject to change, but for the most part I think this is the way to go. I think this very nicely brings us back to solving some of the original problems I was trying to solve with the original system. The reason I had made InputDevice and OutputDevice classes was because I wanted to encapsulate update logic in one spot, rather than force the user to implement that. A major downside of that was tons of extra boilerplate. Here, we've got an overall more terse solution, but it's one the user can opt in to. If they've just got a one-shot effect, or don't need the added features of sliders, time information, etc, then they don't have to use a driver. Or, if they need a super customized solution for their project, there's no reason they couldn't develop their own version of a driver.

**Comment by DavJCosby** [comment]
I've merged the 0.2.0-proto branch into master. Decided that master will hold the latest stable changes as this gets fleshed out, as new features are worked on in other branches and pulled in.

**Comment by DavJCosby** [comment]
I think at this point the architecture has solidified and we've brought this to a conclusion!
